File: python/901. Online Stock Span.py
class StockSpanner(object):

    def __init__(self):
        self.stock = []
        self.span = []

    def next(self, price):
        """
        :type price: int
        :rtype: int
        """
        w = 1
        while self.stock and self.stock[-1] <= price:
            self.stock.pop()
            w += self.span.pop()
        self.stock.append(price)
        self.span.append(w)
        # print self.span, self.stock
        return self.span[-1]




# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: go/src/leetcode/901/901. Online Stock Span.go
/**
@author ZhengHao Lou
Date    2022/10/21
*/
package main

/**
https://leetcode.cn/problems/online-stock-span/
*/
type StockSpanner struct {
	spans  []int
	stocks []int
}

func Constructor() StockSpanner {
	return StockSpanner{}
}

func (this *StockSpanner) Next(price int) int {
	var w = 1
	for len(this.stocks) != 0 && this.stocks[len(this.stocks)-1] <= price {
		this.stocks = this.stocks[:len(this.stocks)-1]
		w += this.spans[len(this.spans)-1]
		this.spans = this.spans[:len(this.spans)-1]
	}
	this.stocks = append(this.stocks, price)
	this.spans = append(this.spans, w)
	return this.spans[len(this.spans)-1]
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */
